discord.go

```diff
@@ -96,35 +98,202 @@ func (dm *DiscordManager) SaveServerOptions(oldServerOpts *GlobalServerConfig, n
+			member, err = s.GuildMember(m.GuildID, m.Author.ID)
+		}
 
 		if err != nil {
-			return
+			return "You don't seem to exist, so I can't assign you that role.  Try again later.", err
 		}
+	}
 
-		var msg = ""
+	// get the role name from ID, for use in user feedback
+	roleName := "notification"
+	roles, err := s.GuildRoles(m.GuildID)
 
-		for _, event := range calendar {
-			msg += event.Start.Format("Mon, 02 Jan 2006 15:04:05 MST") + "\n"
-			msg += event.Title + "\n"
-			msg += event.Description + "\n\n"
+	if err != nil {
+		// meh, just log it and carry on
+		logrus.WithError(err).Infof("failed to get Discord roles (make sure you have set the bot permissions to see roles)")
+	} else {
+		for _, role := range roles {
+			if strings.TrimSpace(role.ID) == strings.TrimSpace(serverOpts.DiscordRoleID) {
+				roleName = role.Name
+				break
+			}
 		}
+	}
 
-		_, err = s.ChannelMessageSend(m.ChannelID, msg)
+	for _, roleID := range member.Roles {
+		if roleID == serverOpts.DiscordRoleID {
+			// they have the role, so remove it
+			err = s.GuildMemberRoleRemove(m.GuildID, m.Author.ID, serverOpts.DiscordRoleID)
 
-		if err != nil {
-			logrus.Errorf("couldn't open discord session, err: %s", err)
+			if err != nil {
+				// meh, log the error here, and just return some feedback to the user
+				logrus.WithError(err).Infof("failed to remove Discord role (make sure you have set the bot permissions to manage roles)")
+				return fmt.Sprintf("You already have the %s role, and an error occurred trying to remove it.  Try again later.", roleName), nil
+			}
+
+			return fmt.Sprintf("You already had the %s role, it has now been removed.  Type the command again to add it back.", roleName), nil
 		}
 	}
+
+	// they didn't have the role, so add it
+	err = s.GuildMemberRoleAdd(m.GuildID, m.Author.ID, serverOpts.DiscordRoleID)
+
+	if err != nil {
+		// meh, log the error here and return feedback to the user
+		logrus.WithError(err).Infof("failed to set Discord role (make sure you have set the bot permissions to manage roles)")
+		return fmt.Sprintf("A server error occurred trying to assign you the %s role, try again later", roleName), nil
+	}
+
+	// w00t!
+	return fmt.Sprintf("The %s role has been assigned, you will now get pinged with notifications.  Type the command again to remove it."), nil
+}
+
+func (dm *DiscordManager) CommandHandler(s *discordgo.Session, m *discordgo.MessageCreate) {
+	serverOpts, err := dm.store.LoadServerOptions()
+
+	if err != nil {
+		logrus.WithError(err).Errorf("couldn't load server opts")
+		return
+	}
+
+	if m.Author.ID == s.State.User.ID {
+		return
+	}
+
+	var msg = ""
+
+	switch m.Content {
+	case "!schedule":
+		msg, err = dm.CommandSchedule()
+	case "!sessions":
+		msg, err = dm.CommandSessions()
+	case "!" + serverOpts.DiscordRoleCommand:
+		msg, err = dm.CommandNotify(s, m)
+	default:
+		return
+	}
+
+	if err != nil {
+		logrus.WithError(err).Errorf("Error during handling of Discord command")
```

Comment:
if the err here is non-nil, do you want the ChannelMessageSend below to occur?

Reply:
Yes, because the specific command handler will have returned some user feedback in msg, like "An error occurred trying to assign you that role, please try again later", rather than just silently failing.  So I log the error for the admin to see, and send the user feedback.